src/main.c

```diff
@@ -324,6 +324,13 @@ int main(int argc, char* argv[]) {
   (void) setlocale(LC_ALL, "");
 #endif
 
+#ifdef __OpenBSD__
+  if (pledge("stdio rpath", NULL) == -1) {
+    perror("pledge");
+    exit(JQ_ERROR_SYSTEM);
```

Comment:
This shoudl be
```suggestion
    ret = JQ_ERROR_SYSTEM;
    goto out;
```

Reply:
That would crash in `jq_util_input_free(&input_state);` since `input_state` is declared and defined after pledge, i.e. undefined in the error case.

Why is `out` needed, anyway?  perror(3) writes to stderr whilst the `badwrite` logic is about stdout.  None of the cleanup should be required the progress exits at a point where only `ARGS` and `program_arguments` are initialised.

Reply:
If `out` should be used, either pledge must be called after `jq_util_input_init()` (which is too late, imho) or `jq_util_input_state` must be initialised before pledge such that its cleanup in pledge's error case will succeed... or `out` needs another check.

All that seems a little too much compared to dead simple `perror(); exit();` for a super early failure case that should never happen, anyway.

If pledge(2) returns -1, it means either
- its arguments are invalid (programmer mistake, this diff is correct), or
- pledges are increaed (cannot be, we pledge once), or
- the argument pointers point outside the address space (won't happen with fixed strings and NULL), or
- whatever cosmic unforseen OS error appears during early jq init, in which case I'd always stick to fail-hard-and-fail-early, anyway

Reply:
> That would crash in `jq_util_input_free(&input_state);` since `input_state` is declared and defined after pledge, i.e. undefined in the error case.

Well, that is a bug too. If `jq_init()` fails, that line will cause a crash for the same reason. `input_state` should be initialised to `NULL`, but it is not.

> Why is `out` needed here, anyway? perror(3) writes to stderr whilst the `badwrite` logic is about stdout. None of the cleanup should be required the progress exits at a point where only `ARGS` and `program_arguments` are initialised.

Indeed to free `ARGS`, and `program_arguments`. The return values of `jv_array()`, and `jv_object()` are heap allocated.

Reply:
> > That would crash in `jq_util_input_free(&input_state);` since `input_state` is declared and defined after pledge, i.e. undefined in the error case.
> 
> Well, that is a bug too. If `jq_init()` fails, that line will cause a crash for the same reason. `input_state` should be initialised to `NULL`, but it is not.

I can try fixing all that if you really want to use `out` here.

> 
> > Why is `out` needed here, anyway? perror(3) writes to stderr whilst the `badwrite` logic is about stdout. None of the cleanup should be required the progress exits at a point where only `ARGS` and `program_arguments` are initialised.
> 
> Indeed to free `ARGS`, and `program_arguments`. The return values of `jv_array()`, and `jv_object()` are heap allocated.

But they do not contain any user data at this point and are cleaned up by the kernel during process teardown, which happens immediately.

Reply:
Either call `pledge(); exit(...)` before initalising `ARGS` and `program_argument`, or use `ret = ...; goto out;` and fix the unitialised `input_state` (which needs to be done anyhow, thank you for spotting it).

Reply:
@klemensn yes, it may be better to `exit()` as soon as pledge fail, but you should create those kinds of leaks for no reason or debug tools will complain when we try to use them. 
Either call `pledge(); exit(...)` before initalising `ARGS` and `program_argument`, or use `ret = ...; goto out;` and fix the unitialised `input_state` (which needs to be done anyhow, thank you for spotting it).

Reply:
Right, might as well defer initialisation for those two variables and call it a day wrt. pledge -- that's much easier and the actual `out`/cleanup bug can be handled separately.

Reply:
I don't see how it is not easier to use `goto out`, but if you insist fine I guess.

Reply:
Using `out` seemed trivial, but uncovered a bug and, imho, means keeping track of a lot more state than just bailing out as early as possible.

Reply:
Yes, I meant, the patch would be neater without reordering the initialisation, but bailing as soon as possible is better, and also might aswell defer the allocations, so fine; let's do it.

Reply:
Meh, I dropped the comments... fixing the fix now!